src/helm/benchmark/presentation/run_display.py

```diff
@@ -0,0 +1,267 @@
+
+
+def _read_per_instance_stats(run_path: str) -> List[PerInstanceStats]:
+    per_instance_stats_path: str = os.path.join(run_path, "per_instance_stats.json")
+    if not os.path.exists(per_instance_stats_path):
+        raise ValueError(f"Could not load PerInstanceStats from {per_instance_stats_path}")
+    with open(per_instance_stats_path) as f:
+        raw_per_instance_stats = json.load(f)
+        return [dacite.from_dict(PerInstanceStats, r, config=_DACITE_CONFIG) for r in raw_per_instance_stats]
+
+
+def _truncate_predicted_text(
+    predicted_text: str, request_state: RequestState, adapter_spec: AdapterSpec
+) -> Optional[str]:
+    method = adapter_spec.method
+    prefix = ""
+    if method.startswith("multiple_choice_separate_"):
+        prefix = request_state.instance.input
+    elif method == "language_modeling":
+        if request_state.result is not None and request_state.result.completions:
+            tokens = request_state.result.completions[0].tokens
+            if tokens:
+                first_token = tokens[0]
+                if not first_token.top_logprobs:
+                    prefix = first_token.text
+    if prefix:
+        predicted_text = predicted_text.strip()
+        prefix = prefix.strip()
+        if predicted_text.startswith(prefix):
+            return predicted_text[len(prefix) :].strip()
+    return None
+
+
+@htrack(None)
+def write_run_display_json(run_path: str, run_spec: RunSpec, schema: Schema):
+    """Write run JSON files that are used by the web frontend.
+
+    The derived JSON files that are used by the web frontend are much more compact than
+    the source JSON files. This speeds up web frontend loading significantly.
+
+    Reads:
+
+    - ScenarioState from `scenario_state.json`
+    - List[PerInstanceStats] from `per_instance_stats.json`
+
+    Writes:
+
+    - List[Instance] to `instances.json`
+    - List[DisplayPrediction] to `predictions.json`
+    - List[DisplayRequest] to `requests.json`
+    """
+    scenario_state = _read_scenario_state(run_path)
+    per_instance_stats = _read_per_instance_stats(run_path)
+
+    metric_groups_by_name = {metric_group.name: metric_group for metric_group in schema.metric_groups}
+    run_groups_by_name = {run_group.name: run_group for run_group in schema.run_groups}
+
+    metric_names_from_schema: Set[str] = set(
+        metric_name_matcher.substitute(run_groups_by_name[run_group_name].environment).name
+        for run_group_name in run_spec.groups
+        if run_group_name in run_groups_by_name
+        for metric_group_name in run_groups_by_name[run_group_name].metric_groups
+        if metric_group_name in metric_groups_by_name
+        for metric_name_matcher in metric_groups_by_name[metric_group_name].metrics
+        if not metric_name_matcher.perturbation_name
+    )
+    if run_spec.adapter_spec.method.startswith("multiple_choice_separate_"):
+        metric_names_from_schema.add("predicted_index")
```

Comment:
Can we pull this out into another function, since this seems like a generalizable unit?

I find this block a bit hard to read...

Reply:
Rewrote this, but unfortunately we still need deep nesting because the MetricNameMatcher environment substitution requires the run from the outermost loop.